### PR TITLE
Refine STO consensus comparison and fee tests

### PR DESCRIPTION
**Pull request by dexX7**

### CODE REVIEW

**Comment by dexX7** [comment]
Actually there is one case still left out. Solely running MSCSendToOwnersConsensusComparisonSpec as a fresh start causes a failure, because without any other holder there can't be a complete STO, thus no change in balance. Needs something like [.../MSCSendToOwnersSpec.groovy#L45-L55](https://github.com/dexX7/bitcoin-spock/blob/sto-consensus-comparison/src/integ/groovy/org/mastercoin/test/rpc/MSCSendToOwnersSpec.groovy#L45-L55) as well. I'll append this part tomorrow.

File: src/integ/groovy/org/mastercoin/test/rpc/MSCSendToOwnersSpec.groovy [hunk 1]
```diff
@@ -27,21 +27,32 @@ class MSCSendToOwnersSpec extends BaseRegTestSpec {
...
-        and: "we generate a block"
+        and: "We generate a block"
         generateBlock()
-        def balances = getallbalancesforid_MP(currencyID)
-        def numberDest = balances.size() - 1
-        def totalFee = numberDest * stoFeePerAddress
-        def expectedBalance = (numberDest == 0) ? startingPropBal : startingPropBal - amountSent - totalFee
+
+        // The fee for each receiver is #stoFeePerAddress
+        if (numberOfHolders > 1) {
```
Comment (msgilligan) [inline src/integ/groovy/org/mastercoin/test/rpc/MSCSendToOwnersSpec.groovy]: What happens if the number of holders changes within this block? Are there cases we might want to test where the results could be dependent on transaction order?

Comment (dexX7) [inline src/integ/groovy/org/mastercoin/test/rpc/MSCSendToOwnersSpec.groovy]: Good point. In regtest mode we have full control here, but this isn't necessarily the case on mainnet. What do you suggest? The RPC call `getsto_MP` has a fee field, but it introduces another layer and there was already one flaw in the fee calculation in Core, see https://github.com/mastercoin-MSC/mastercore/pull/271.

Comment (msgilligan) [inline src/integ/groovy/org/mastercoin/test/rpc/MSCSendToOwnersSpec.groovy]: I don't know the spec or the implementation as well as you, but what I was thinking there might be a case where an address receives a SP that it didn't previously own in the same block that an STO was being made for that token. Wouldn't the behavior depend upon whether the transaction adding the SP to a new address occurred before or after the STO?

I realize we (likely) have control over this in RegTest mode so I'm thinking we could write tests to verify that this is being handled correctly in the Core implementation.

Comment (dexX7) [inline src/integ/groovy/org/mastercoin/test/rpc/MSCSendToOwnersSpec.groovy]: This is something that made me wonder at some point, too. But it's less complicated than expected: transactions are parsed and processed in order of their appearance and position within a block. If an address receives additional token before the STO transaction is processed, then the "send to owners" will use that updated value.

Comment (msgilligan) [inline src/integ/groovy/org/mastercoin/test/rpc/MSCSendToOwnersSpec.groovy]: Right. So shouldn't we write some tests to verify/specify that behavior?

Comment (dexX7) [inline src/integ/groovy/org/mastercoin/test/rpc/MSCSendToOwnersSpec.groovy]: Sure, I think this would be beneficial, but probably in form of a different PR. The intend of this one was not necessarily to change the logic, but fix the issue of test failures for the two cases:

- There are many holders and the amount sent to the owners doesn't "reach" everyone. The unreached addresses create a failure. This is why I introduced the "number of owners" and "balances changed".
- At the time of test start there is no holder other than the one we create. In this case the test fails, because the sender's balance doesn't change at all. It's addressed in MSCSendToOwnersSpec, but not yet in MSCSendToOwnersConsensusComparisonSpec.

Comment (msgilligan) [inline src/integ/groovy/org/mastercoin/test/rpc/MSCSendToOwnersSpec.groovy]: Agreed that those tests would be for a different PR.
